feat: use ko to build images

imjasonh opened the pull request:
## Explanation

Instead of running `docker build` to execute Dockerfiles that amount to `go build`ing and putting the resulting binary on top of a distroless base, use `ko`, which is specifically designed and optimized for exactly this case.

Using `ko` has some advantages:
- it reuses your Go build cache by default
- it doesn't require maintaining a `Dockerfile` or executing `docker build`
- it generates SBOMs by default
- it doesn't require qemu to build multi-platform images (including Windows images!)
- probably more, I forget 😁 

## Related issue

https://github.com/kyverno/kyverno/issues/2925

## Milestone of this PR

TBD

## What type of PR is this

/kind feature

## Proposed Changes

Remove `Dockerfile`s, replace invocations of `docker buildx build` in Makefile with `ko build`, and clean up anything else I can find along the way.

### Proof Manifests

N/A

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [ ] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the documentation update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

I've started with just one image, `initContainer`, to prove the process. Once that looks good I can extend to the other images, or do that in separate PRs if there's a preference.

cc @mattmoor @chipzoller @JimBugwadia @developer-guy

imjasonh commented:
With the latest commit the image name should be equivalent, and have all the same tags applied:

```
$ REGISTRY=gcr.io/jason-chainguard make ko-build-initContainer
...
gcr.io/jason-chainguard/kyverno/kyvernopre:v1.5.0-rc1-810-g9be68cbb1@sha256:d80a87d99f88b8f600ed7f9e582d2e8e24ed93e07272b12c55151c7a529dac06
```

eddycharly commented:
Thanks @imjasonh for working on this !

`make: ko: Command not found` in e2e tests.

imjasonh commented:
> Thanks @imjasonh for working on this !
> 
> 
> 
> `make: ko: Command not found` in e2e tests.

Oh yep I'll need to add `setup-ko` for e2e tests too.

eddycharly commented on Makefile:
```diff
@@ -61,40 +61,37 @@ INITC_IMAGE := kyvernopre
 	fi
 
-docker-publish-initContainer: docker-buildx-builder docker-build-initContainer docker-push-initContainer
+# TODO(jason): LDFLAGS
```
I was going to ask about `LDFLAGS` ;)

developer-guy replied:
There will be a configuration file named `.ko.yaml`, for example, you can look at ko's default configuration [file](https://github.com/google/ko/blob/main/.ko.yaml#L7). Then, we pass these LDFLAGS to that file via the environment variable.

imjasonh replied:
Yep I meant to do that yesterday but just didn't have time before my son woke up from his nap 😆 

This should be done now, at least for `LD_FLAGS` -- I'll need to up my Makefile-fu to pass in the correct `LD_FLAGS_DEV`, then I can mark this as ready for review.

imjasonh commented:
```
REGISTRY=gcr.io/jason-chainguard-public make ko-build-all
...
gcr.io/jason-chainguard-public/kyverno/kyverno-cli:v1.5.0-rc1-811-g244985b19@sha256:891d1a7cbc9bc13398aa66dc6ff71fbd8c4cdcc1d517698c3b3b1e8deaf56b1c
```

You can inspect the images this produces here: https://console.cloud.google.com/gcr/images/jason-chainguard-public/global/kyverno

eddycharly commented:
@imjasonh would it make sense to install ko in the makefile ?

imjasonh commented:
Some things I noticed that we might want to consider addressing in followup PRs:
- The hard-coded image reference in [`install.yaml`](https://github.com/kyverno/kyverno/blob/8e5f831b075d8b5cea9e411c03e11a89fa88f029/config/release/install.yaml#L7727) could be replaced by `ko://`-prefixed references, and generated using `ko resolve -f install.yaml` to build the images and fill in the placeholders with image references by digest. This is a bit safer than relying on the tag not to change, but it would mean folks can't `kubectl apply -f install.yaml` directly from the repo, they'd need to either `ko apply` it (building the images themselves) or `kubectl apply -f release.yaml` built from some other release workflow. I'm not sure if that's the intended use of `install.yaml` though.
 
- The image release process in `reuse.yaml` generates another SBOM based on the contents of `go.mod` -- this can be misleading in some cases, if dependencies listed in `go.mod` aren't present in the binary in the final built image. This can happen for example for testing dependencies like `github.com/stretchr/testify`. `ko`'s SBOMs are generated from the output of `go version -m <binary>` which only includes the dependencies actually included at build-time.

imjasonh commented:
> @imjasonh would it make sense to install ko in the makefile ?

Sure, let me work on that a bit.

eddycharly commented:
Thanks for sharing potential improvement ideas, we can address them in follow up PRs.
Would be awesome to create issues to be sure we don't forget about it though.

eddycharly commented:
Looks like there's an issue with CLI and LDFLAGS:
`flag provided but not defined: -ldflags-s -w -X github.com/kyverno/kyverno/pkg/version.BuildVersion`

eddycharly commented:
Loading images directly into kind broke chart testing (chart values use `ghcr.io/kyverno/...`)

eddycharly commented on Makefile:
```diff
@@ -327,7 +279,9 @@ helm-test-values:
 	sed -i -e "s|namespace:.*|namespace: kyverno|g" charts/kyverno/values.yaml
-	sed -i -e "s|tag:  # replaced in e2e tests.*|tag: $(GIT_VERSION_DEV)|" charts/kyverno/values.yaml
+	sed -i -e "s|tag:  # replaced in e2e tests.*|tag: $(IMAGE_TAG_DEV)|" charts/kyverno/values.yaml
+	sed -i -e "s|repository: ghcr.io/kyverno/kyverno  # kyverno: replaced in e2e tests|repository: $(KYVERNO_KIND_IMAGE)|" charts/kyverno/values.yaml
+	sed -i -e "s|repository: ghcr.io/kyverno/kyvernopre  # init: replaced in e2e tests|repository: $(INITC_KIND_IMAGE)|" charts/kyverno/values.yaml
```
```suggestion
	sed -i -e "s|repository: ghcr.io/kyverno/kyverno|repository: $(KYVERNO_KIND_IMAGE)|" charts/kyverno/values.yaml
	sed -i -e "s|repository: ghcr.io/kyverno/kyvernopre|repository: $(INITC_KIND_IMAGE)|" charts/kyverno/values.yaml
```

eddycharly replied:
Substituting `kyvernopre` should occur before substituting `kyverno` though.

eddycharly commented:
@imjasonh if you keep the comments in the see commands you will need to add them in values.yaml (else the command won’t match)

imjasonh commented:
> @imjasonh if you keep the comments in the `sed` commands (`sed -i -e "s|... # init: replaced in e2e tests|...|"`) you will need to add the same comments in `values.yaml` too (else the command won’t match)

D'oh! I had that but I reverted `values.yaml` before pushing because it included generated contents. Pushed with a fix. 🤞

eddycharly commented:
Thanks for the work @imjasonh !
We are releasing 1.8 soon, any risk merging this before branching ? (any chance it will break the release process ?)

imjasonh commented:
> Thanks for the work @imjasonh ! We are releasing 1.8 soon, any risk merging this before branching ? (any chance it will break the release process ?)

I would probably wait until after a release. I don't know enough about the current release process to know whether I've made all the necessary changes.

I can run some test releases in my fork in the meantime to try to flush out anything else I'm missing.

eddycharly commented:
IMHO we should discuss it with the team before merging.
Every contributor/maintainer will need to install ko and potentially adapt the way they work.
We probably need to update a couple of docs too.

Regarding the release process, we start with a couple of pre releases, we can use them to adjust if necessary.

I will sync with the team tomorrow.

imjasonh commented:
Absolutely. Let me know if I can be helpful, I'm happy to join any meetings or slack conversations, and I can make any changes to the PR based on feedback.

Thanks for your help so far! 🙏

eddycharly commented:
We have a contributors meeting every Wednesday at 6PM (CEST). Could be a good opportunity to discuss it with the team.

eddycharly commented:
> There's a bit of complexity installing tools for different OS environments

Can't we just use `go install` ?

imjasonh commented:
> > There's a bit of complexity installing tools for different OS environments
> 
> Can't we use `go install` ?

Sure, I see you're already doing that with kind, so this should be pretty much the same.

developer-guy commented on .github/workflows/reuse.yaml:
```diff
@@ -152,7 +135,7 @@ jobs:
           -a "workflow=${{ github.workflow }}" \
           -a "ref=${{ github.sha }}" \
-          ghcr.io/${{ github.repository_owner }}/${{inputs.image_name}}@sha256:${{ steps.get-step-image.outputs.digest }}
+          ${{ steps.ko-build-dev.outputs.digest }}
```
shouldn't it be "ko-build" instead of "ko-build-dev"?

imjasonh replied:
Close! The `ko build dev image` step should have had `id: ko-build-dev` so that the `Sign dev image` can use its output.

I'm surprised GitHub let me define two steps with the same `id` but 🤷 -- this just points out that we'll need to do more testing of this release workflow before we can merge this PR.

eddycharly commented:
I just encountered issues with `go install`.
Probably not a good idea to install `ko` this way.

We can address it later.

imjasonh commented:
> I just encountered issues with `go install`. Probably not a good idea to install `ko` this way.
> 
> We can address it later.

What issues are you hitting?

eddycharly commented:
I used `go install` to install `golangci-lint` but it didn’t behave the same as installing pre-built binary.

Probably because of a different go version used for compiling, different version of a standard lib, something like this…

eddycharly commented:
Another difference is that additional build args are not passed to the build (`ldflags` for example).

imjasonh commented:
> Another difference is that additional build args are not passed to the build (`ldflags` for example).

ldflags _should_ still be passed through, unless I'm missing something. I think there's still a TODO about LDFLAGS_DEV, but I can address that before merging this.

imjasonh commented:
Something I noticed while running through a release: Currently the release process kicks off three separate workflows to build kyvernopre, kyverno, and the CLI, and waits for all three before creating the release.

I think we could make this go a bit faster by having all three images built in the same workflow (e.g., using `make ko-build-all`), so they share the same build cache and can share built dependencies.

This could even be something worth doing with Dockerfile-based builds, if we decide not to go with `ko`. But it could also happen in a followup PR, since it's just an optimization and not a behavior change.

eddycharly commented:
> ldflags should still be passed through, unless I'm missing something

I was talking about `go install` here, `go install` builds from source and therefore additional build args are not passed, that can explain why a tool installed with `go install` has good chances to be different from a pre-built binary.

imjasonh commented:
> > ldflags should still be passed through, unless I'm missing something
> 
> I was talking about `go install` here, `go install` builds from source and therefore additional build args are not passed, that can explain why a tool installed with `go install` has good chances to be different from a pre-built binary.

Oh I see. Ideally `ko` wouldn't rely on ldflags to be able to print `ko version`, since it can just get that information from [`BuildInfo`](https://pkg.go.dev/debug/buildinfo). I can make some changes in `ko`, but I think that's probably not worth blocking on. You can still use `ko` without it, it doesn't add much otherwise.

Instead of installing `ko` if it's not found, we could also have a `check-ko` target that fails with a helpful error message if `ko` isn't installed.

eddycharly commented:
Really not a blocker, I started cleaning up / refactoring our `Makefile` in #4403 , I'm just trying to picture how `ko` will come in once it's refactored.

Don't worry about it, assuming `ko` is present locally and installing in GH workflows for CI is fine.

eddycharly commented:
@imjasonh can you resolve conflict and we merge it ?

eddycharly reviewed (approved):
Thanks, great work !

realshuting commented on DEVELOPMENT.md:
````diff
@@ -0,0 +1,68 @@
+Set the `KO_DOCKER_REPO` environment variable to `ko.local`:
+
+```
+KO_DOCKER_REPO=ko.local
````
Need to run `export KO_DOCKER_REPO=ko.local` on MAC.

realshuting commented:
Excellent work @imjasonh !
